README
-*- encoding: utf-8; indent-tabs-mode: nil -*-

Date::Convert 0.17 pre-alpha by Mordechai T. Abzug
and Jean Forget <JFORGET at cpan dot org>

Allows you to convert date formats using an OO mechanism that lets you
easily choose any two formats and add in new ones. Example:

  use Date::Convert;
  my $today = Date::Convert::Gregorian->new(2020, 10, 12);
  print "Gregorian: ", $today->date_string, "\n";
  # ---> Gregorian: 2020 Oct 12

  Date::Convert::Julian->convert($today);
  print "Julian: ", $today->date_string, "\n";
  # ---> Julian: 2020 Sep 29

  Date::Convert::Hebrew->convert($today);
  print "Hebrew: ", $today->date_string, "\n";
  # ---> Hebrew: 5781 Tishrei 24

  Date::Convert::Absolute->convert($today);
  print "Julian Day Number at noon: ", $today->date_string, "\n";
  # ---> Julian Day Number at noon: 2459135

At the moment, the only documentation is the man page.  That's not
surprising, since it *is* pre-alpha.  And of course, since it's pre-alpha,
'tis subject to change, and is unreliable.

Install the normal way, using cpanminus or the CPAN shell, or by running
the usual commands:

  perl Makefile.PL
  make
  make test
  make install

If you have suggestions, bug reports, or if you want to add a new date
format, feel free to contact me:  morty@sanctuary.arbutus.md.us

If you want to be on the announcement list for new releases, send a mail
with "subscribe" as the first line of the body to:
dateconvert-announce-request@sanctuary.arbutus.md.us



CHANGES
-*- encoding: utf-8; indent-tabs-mode: nil -*-

0.17:   released on xxx
        - cleaner distribution architecture, with a "/lib" subdirectory, MANIFEST, META.json, etc
        - more modern syntax, for example replacing indirect object method calls by arrow method calls
        - use strict and use warnings in all modules
        - fix off-by-400-years or off-by-4-years errors when converting a Hebrew date to a BC Gregorian / Julian date (RT 130291)
        - fix "impossible year-length" errors and off-by-one-day errors in Hebrew dates (RT 98287 and RT 130291)
        - month 12 in leap years is "Adar I" instead of just "Adar" (RT 98287)
        - fix some errors related to Julian 2100-02-29 leap day (RT 98287)
        - fix errors when a date object is built with empty input parameters
        - fix the POD documentation
        - make Date::Convert an abstract class, no instantiation

0.16:
        - BUG: Hebrew broken, off-by-one error in rosh calulations

0.15:
        - BUG: no initialize for Julian!  How embarassing.
        - reworded docs to replace "Absolute" for "Astro"

0.14:
        - first public release


Makefile.PL
#
#     Configuration script for Date::Convert
#     Copyright © 1997, 2000, 2020 Mordechai Abzug and Jean Forget
#
#     This program is distributed under the same terms as Perl 5.16.3:
#     GNU Public License version 1 or later and Perl Artistic License
#
#     You can find the text of the licenses in the F<LICENSE> file or at
#     L<https://dev.perl.org/licenses/artistic.html>
#     and L<https://www.gnu.org/licenses/gpl-1.0.html>.
#
#     Here is the summary of GPL:
#
#     This program is free software; you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation; either version 1, or (at your option)
#     any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program; if not, write to the Free Software Foundation,
#     Inc., <https://www.fsf.org/>.
#
use ExtUtils::MakeMaker;
WriteMakefile(
    'NAME'             => 'Date::Convert'
  , 'VERSION_FROM'     => 'lib/Date/Convert.pm'
  , 'ABSTRACT'         => 'Date conversions between Gregorian, Hebrew and Julian'
  , 'AUTHOR'           => [ 'Mordechai Abzug', 'Jean Forget <JFORGET@cpan.org>' ],
  , 'LICENSE'          => 'perl'
  , 'MIN_PERL_VERSION' => '5.6.0'
  , 'BUILD_REQUIRES'   => { 'Test::More'      => '0'
                          , 'ExtUtils::MakeMaker' => '6.57_02'
                           },
   , META_MERGE       => {
       'meta-spec'      => { version => 2 },
       'dynamic_config' => 0,
       'prereqs' => {
          'build' => {
             'requires' => {
                'ExtUtils::MakeMaker' => '6.57_02'
             }
          },
          'configure' => {
             'requires' => {
                'ExtUtils::MakeMaker' => '6.57_02'
             }
          },
          'runtime' => {
             'requires' => {
                 'POSIX'           => '0'
               , 'strict'          => '0'
               , 'warnings'        => '0'
               , 'Carp'            => '0'
               , 'perl'            => '5.006000'
             }
          },
          'test' => {
             'recommends' => {
                  'Test::Exception' => '0'
             }
          },
       },
       provides       => {
           'Date::Convert' => {
                file => 'lib/Date/Convert.pm',
                version => '0.17',
           },
           'Date::Convert::Absolute' => {
                file => 'lib/Date/Convert/Absolute.pm',
                version => '0.17',
           },
           'Date::Convert::Gregorian' => {
                file => 'lib/Date/Convert/Gregorian.pm',
                version => '0.17',
           },
           'Date::Convert::Hebrew' => {
                file => 'lib/Date/Convert/Hebrew.pm',
                version => '0.17',
           },
           'Date::Convert::Julian' => {
                file => 'lib/Date/Convert/Julian.pm',
                version => '0.17',
           },
       },
       resources => {
           repository => {
               type => 'git',
               url  => 'https://github.com/jforget/Date-Convert.git',
               web  => 'https://github.com/jforget/Date-Convert',
           },
       },
    },
);



lib/Date/Convert/Hebrew.pm
# -*- encoding: utf-8; indent-tabs-mode: nil -*-
#
#     Utility module to deal with Hebrew dates in Date::Convert
#     Copyright (c) 1997, 2000, 2020 Mordechai Abzug and Jean Forget
#
#     See the license in the embedded documentation below.
#

package Date::Convert::Hebrew;
use utf8;
use strict;
use warnings;
use Carp;

our @ISA     = qw ( Date::Convert );
our $VERSION = "0.17";

my $HEBREW_BEGINNING = 347996; # 1 Tishri 1
#  @MONTH       = (  29, 12, 793);
my @NORMAL_YEAR = ( 354,  8, 876); # _part_mult(12,  @MONTH);
my @LEAP_YEAR   = ( 383, 21, 589); # _part_mult(13,  @MONTH);
my @CYCLE_YEARS = (6939, 16, 595); # _part_mult(235, @MONTH);
my @FIRST_MOLAD = (   1,  5, 204);

#
# Which years in the Metonic cycle are leap years?
#
my @LEAP_BOOL   = qw ( 1 0 0 1 0 0 1 0 1 0 0 1 0 0 1 0 0 1 0 );
my @LEAP_CYCLE  = qw (       3     6   8    11    14    17
                       0 );

my @MONTHS = ('Nissan',  'Iyyar',    'Sivan',  'Tammuz', 'Av',     'Elul',
              'Tishrei', 'Cheshvan', 'Kislev', 'Teves',  'Shevat', 'Adar', 'Adar II' );

# In the Hebrew calendar, the year starts in the seventh month, there can
# be a leap month, and there are two months with a variable number of days.
# Rather than calculate do the actual math, let's set up lookup tables based
# on year length.  :)

my %MONTH_START=
    ('353' => [177, 207, 236, 266, 295, 325, 1, 31, 60, 89, 118, 148],
     '354' => [178, 208, 237, 267, 296, 326, 1, 31, 60, 90, 119, 149],
     '355' => [179, 209, 238, 268, 297, 327, 1, 31, 61, 91, 120, 150],
     '383' => [207, 237, 266, 296, 325, 355, 1, 31, 60, 89, 118, 148, 178],
     '384' => [208, 238, 267, 297, 326, 356, 1, 31, 60, 90, 119, 149, 179],
     '385' => [209, 239, 268, 298, 327, 357, 1, 31, 61, 91, 120, 150, 180]);

my %MONTH_LENGTH =
    ('353' => [30, 29, 30, 29, 30, 29, 30, 29, 29, 29, 30, 29],
     '354' => [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 29],
     '355' => [30, 29, 30, 29, 30, 29, 30, 30, 30, 29, 30, 29],
     '383' => [30, 29, 30, 29, 30, 29, 30, 29, 29, 29, 30, 30, 29],
     '384' => [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 30, 29],
     '385' => [30, 29, 30, 29, 30, 29, 30, 30, 30, 29, 30, 30, 29]);

sub is_leap {
    my $self = shift;
    my $year = shift;
    $year = $self->year if ! defined $year;
    my $mod = $year % 19;
    return $LEAP_BOOL[$mod];
}


sub initialize {
  my ($self, $year, $month, $day) = @_;
  return $self->Date::Convert::initialize
    unless defined($year);
  croak "Date::Convert::Gregorian::initialize needs more args"
    unless defined($month)
       and defined($day);

  # coarse check for month number and day number, it will be refined later.
  croak "month number $month out of range"
    if $month < 1 || $month > 13;
  croak "day number $day out of range for month $month"
    if   $day < 1 || $day > 30;

  warn "These routines don't work well for Hebrew before year 1"
    if $year<1;
  $self->{year}   = $year;
  $self->{$month} = $month;
  $self->{day}    = $day;
  my $rosh = $self->rosh;
  my $year_length = Date::Convert::Hebrew->rosh($year+1) - $rosh;

  # If this error is triggered, that means there is a bug in Date::Convert::Hebrew.
  # It does not mean that the user sent input parameters with invalid values.
  carp "Impossible year length $year_length"
    unless defined $MONTH_START{$year_length};

  my $months_ref = $MONTH_START{$year_length};

  # refined check for month number and day number
  croak "month number $month out of range"
    if $month > scalar @$months_ref;
  croak "day number $day out of range for month $month"
    if   $day > $MONTH_LENGTH{$year_length}[$month - 1];

  my $days  = $months_ref->[$month-1] + $day - 1;
  my $absol = $rosh + $days - 1;
  $self->{days}  = $days;
  $self->{absol} = $absol;
}

sub year {
  my $self = shift;

  # no point recalculating.
  return $self->{year}
    if exists $self->{year};

  my $days = $self->absol;
  my $year = int($days / 365) - 3 * 365; # just an initial guess, but a good one.
  $year = 0
    if $year < 0;
  warn "Date::Convert::Hebrew isn't reliable before the beginning of\n\tthe Hebrew calendar"
    if $days < $HEBREW_BEGINNING;

  $year++
    while Date::Convert::Hebrew->rosh($year+1) <= $days;
  $self->{year} = $year;
  $self->{days} = $days - Date::Convert::Hebrew->rosh($year) + 1;
  return $year;
}

sub month {
  my $self = shift;

  # no point recalculating.
  return $self->{month}
    if exists $self->{month};

  my $year_length =   Date::Convert::Hebrew->rosh($self->year+1)
                    - Date::Convert::Hebrew->rosh($self->year);

  # If this error is triggered, that means there is a bug in Date::Convert::Hebrew.
  # It does not mean that the user sent input parameters with invalid values.
  carp "Impossible year length $year_length"
    unless defined $MONTH_START{$year_length};

  my $months_ref  = $MONTH_START{$year_length};
  my $days        = $self->{days};
  my ($n, $month) = (1);
  my $day         = 31; # 31 is too large.  Good.  :)

  for (@$months_ref) {
    if ($days>=$_ && $days-$_<$day) {
      $day   = $days - $_ + 1;
      $month = $n
    }
    $n++;
  }
  $self->{month} = $month;
  $self->{day}   = $day;
  return $month;
}

sub day {
  my $self = shift;

  # no point recalculating.
  return $self->{day}
    if exists $self->{day};

    $self->month; # calculates day as a side-effect.
    return $$self{day};
}


sub date {
    my $self = shift;
    return ($self->year, $self->month, $self->day);
}


sub date_string {
  my $self = shift;
  my $month_name = $MONTHS[$self->month - 1];
  if ($self->month == 12 and $self->is_leap) {
    $month_name = 'Adar I';
  }
  return join ' ', $self->year, $month_name, $self->day;
}


sub rosh {
  my ($self, $year) = @_;
  $year = $self->year
    unless defined($year);

    my @molad= @FIRST_MOLAD;
    @molad = _part_add(@molad, _part_mult(int(($year-1) / 19), @CYCLE_YEARS));
    my $offset=($year-1)%19;
    my $num_leaps=(grep {$_<=$offset} @LEAP_CYCLE) - 1;
    @molad = _part_add(@molad, _part_mult($num_leaps          , @LEAP_YEAR));
    @molad = _part_add(@molad, _part_mult($offset - $num_leaps, @NORMAL_YEAR));
    my $day=shift @molad;
    my $hour=shift @molad;
    my $part= shift @molad;
    my $guess=$day%7;
    if (($hour>=18)                              # molad zoken al tidrosh
	or
	((! Date::Convert::Hebrew->is_leap($year)) and   # gatrad b'shanah
	 ($guess==2) and                         #      p'shutah g'rosh
	 (($hour>9)or($hour==9 && $part>=204)))
        or
        ((is_leap Date::Convert::Hebrew $year-1) and # b'to takfat achar
	 ($guess==1) and                         #      ha'ibur akor
	 (($hour>15)or($hour==15&&$part>589)))){ #      mi-lishorsh
	$guess++;
	$day++;
    }
    $guess%=7;
    if (scalar(grep {$guess==$_} (0, 3, 5))) {   # lo ad"o rosh
	$guess++;
	$day++;
    }
    $guess%=7;
    return ($day+1+$HEBREW_BEGINNING);
}

sub _part_add {
    my ($day1, $hour1, $part1)=(shift, shift, shift);
    my ($day2, $hour2, $part2)=(shift, shift, shift);
    my $part=$part1+$part2;
    my $hour=$hour1+$hour2;
    my $day =$day1 +$day2;
    if ($part>1080) {
	$part-=1080;
	$hour++;
    }
    if ($hour>24) {
	$hour-=24;
	$day++;
    }
    return ($day, $hour, $part);
}


sub _part_mult {
    my $scalar = shift;
    my $day= ((0+ shift) * $scalar);
    my $hour=((0+ shift) * $scalar);
    my $part=((0+ shift) * $scalar);
    my $tmp;
    if ($part>1080) {
	$tmp=int($part/1080);
	$part%=1080;
	$hour+=$tmp;
    }
    if ($hour>24) {
	$tmp=int($hour/24);
	$hour%=24;
	$day+=$tmp;
    }
    return($day, $hour, $part);
}

# Instead of the usual boring "1" end-of-package value,
# just celebrate the Dumas-esque achievement of releasing
# version 0.17 in 2020, after version 0.16 in 2000
'Vingt ans après...';

__END__

=head1 NAME

Date::Convert::Hebrew - Utility module to deal with Hebrew dates in Date::Convert

=head1 DESCRIPTION

See the full documentation in the main module L<Date::Convert>.

=head1 METHODS

=head2 C<new>

Create a  new object.  You must  provide three  positional parameters:
year, month and day.

=head2 C<convert>

Use an  existing C<Date::Convert::>I<xxx>  object and convert  it into
the C<Date::Convert::Hebrew> class.

=head2 C<year>, C<month>, C<day>, C<absol>, C<date_string>

Accessors,  just  like  the  base  class  C<Date::Convert>  and  other
classes.

=head2 C<date>

Accessor, gives a triplet of values, with year, month and day.

=head2 C<rosh>

Accessor,  gives  the  abolute  value  (JDN)  for  Rosh  HaShana,  the
beginning of the year. It can be used as a method:

  my $jdn = $hebrewdate->rosh;

or as a function:

  my $jdn = Date::Convert::Hebrew->rosh($year);

=head2 C<is_leap>

Accessor,  gives  a boolean which indicates if the year has 12 or 13 months
(0 for a 12-month year, 1 for a 13-month year). It can be used as a method:

  my $flag = $hebrewdate->is_leap;

or as a function:

  my $flag = Date::Convert::Hebrew->is_leap($year);

=head1 NOTES ON HEBREW CALENDAR

This is the traditional Jewish calendar. It's based on the solar year,
on the  lunar month, and  on a number  of additional rules  created by
Rabbis to make life tough on people who calculate calendars. :) If you
actually wade  through the  source, you should  note that  the seventh
month really does come before the first month, that's not a bug.

=head1 SEE ALSO

perl(1)

L<DateTime>, with L<DateTime::Calendar::Hebrew>

L<Date::Converter>

L<Date::Hebrew::Simple>

=head1 AUTHORS

Module creator: Mordechai T. Abzug <morty at umbc dot edu>

Unofficial co-maintainer: Jean Forget <JFORGET at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (c)  1997, 2000, 2020  Mordechai Abzug and Jean  Forget. All
rights reserved.

This  program  is  free  software. You  can  distribute,  modify,  and
otherwise mangle  Date::Convert::Hebrew under  the same terms  as Perl
5.16.3:  GNU Public  License  version  1 or  later  and Perl  Artistic
License

You can  find the text  of the licenses  in the F<LICENSE> file  or at
L<https://dev.perl.org/licenses/artistic.html>
and L<https://www.gnu.org/licenses/gpl-1.0.html>.

Here is the summary of GPL:

This program is  free software; you can redistribute  it and/or modify
it under the  terms of the GNU General Public  License as published by
the Free  Software Foundation; either  version 1, or (at  your option)
any later version.

This program  is distributed in the  hope that it will  be useful, but
WITHOUT   ANY  WARRANTY;   without  even   the  implied   warranty  of
MERCHANTABILITY  or FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
General Public License for more details.

You  should have received  a copy  of the  GNU General  Public License
along with  this program; if not,  see <https://www.gnu.org/licenses/>
or write to the Free Software Foundation, Inc., L<https://www.fsf.org>.

=cut


lib/Date/Convert.pm
# -*- encoding: utf-8; indent-tabs-mode: nil -*-
#
#     Perl module to convert dates between Gregorian, Julian and Hebrew
#     Copyright © 1997, 2000, 2020 Mordechai Abzug and Jean Forget
#
#     See the license in the embedded documentation below.
#

package Date::Convert;

use utf8;
use strict;
use warnings;
use Carp;
use Date::Convert::Absolute;
use Date::Convert::Gregorian;
use Date::Convert::Hebrew;
use Date::Convert::Julian;

our $VERSION = "0.17";

# methods that every class should have:
# initialize, day, date, date_string

# methods that are recommended if applicable:
# year, month, day, is_leap

my $BEGINNING = 1721426; # 1 Jan 1 in the Gregorian calendar, although technically,
                         # the Gregorian calendar didn't exist at the time.
my $VERSION_TODAY = 2450522; # today in JDN, when I wrote this.


sub new { # straight out of the perlobj manpage:
  my $class = shift;
  croak "$class is an abstract class and cannot be instantiated"
    if $class eq "Date::Convert";
  my $self = {};
  bless $self, $class;
  $self->initialize(@_);
  return $self;
}


sub initialize {
  my ($self, $val) = @_;
  $val  = $VERSION_TODAY
    unless defined($val);
  carp "Date::Convert is not reliable before Absolute $BEGINNING"
      if $val < $BEGINNING;
  $self->{absol} = $val;
}


sub _clean {
  my $self  = shift;
  my $key;
  foreach $key (keys %$self) {
    delete $self->{$key} unless $key eq 'absol';
  }
}



sub convert {
    my $class = shift;
    my $self  = shift;
    $self->_clean;
    bless $self, $class;
}

sub absol {
    my $self = shift;
    return $$self{absol};
}

# Instead of the usual boring "1" end-of-package value,
# just celebrate the Dumas-esque achievement of releasing
# version 0.17 in 2020, after version 0.16 in 2000
'Vingt ans après...';

__END__

=head1 NAME

Date::Convert - Convert Between Calendrical Formats

=head1 SYNOPSIS

  use Date::Convert;

  my $date = Date::Convert::Gregorian->new(1997, 11, 27);
  @date    = $date->date;
  Date::Convert::Hebrew->convert($date);
  print $date->date_string, "\n";

=head1 DESCRIPTION

Date::Convert is intended to allow you to convert back and forth between
any arbitrary date formats (i.e. pick any from: Gregorian, Julian, Hebrew,
Absolute, and any others that get added on).  It does this by having a
separate subclass for each format, and requiring each class to provide
standardized methods for converting to and from the date format of the base
class.  In this way, instead of having to code a conversion routine for
going between and two arbitrary formats I<foo> and I<bar>, the function only
needs to convert I<foo> to the base class and the base class to I<bar>.
For example:

  Gregorian <--> Base class <--> Hebrew

The base class includes a B<Convert> method to do this transparently.

Currently defined subclasses:

  Date::Convert::Absolute
  Date::Convert::Gregorian
  Date::Convert::Hebrew
  Date::Convert::Julian

Nothing is exported because it wouldn't make any sense to export.


Functions can be split into several categories:

=over 4

=item *

Universal functions available for all subclasses (ie. all formats).  The
fundamental conversion routines fit this category.

=item *

Functions that are useful but don't necessarily make sense for all
subclasses.  The overwhelming majority of functions fall into this
category.  Even such seemingly universal concepts as year, for instance,
don't apply to all date formats.

=item *

Private functions that are required of all subclasses, ie. B<initialize>.
These should I<not> be called by users.

=back

Here's the breakdown by category:

=head2 Functions Defined for all Subclasses

=over 4

=item new

Create a new object in the specified format with the specified start
parameters, i.e. C<< $date = Date::Convert::Gregorian->new(1974, 11, 27) >>.  The
start parameters vary with the subclass.  My personal preference is to
order in decreasing order of generality (ie. year first, then month, then
day, or year then week, etc.)

This can have a default date, which should probably be "today".

=item date

Extract the date in a format appropriate for the subclass.  Preferably this
should match the format used with B<new>, so

  Date::Convert::SomeClass->new(@a)->date;

should be an identity function on @a if @a is a legitimate input value.

=item date_string

Return the date in a pretty, human-readable, format.

=item convert

Change the date to a new format.

=item absol

Give the absolute  value of the date, that is,  the Julian Day Number,
at noon.

=back

=head2 Non-universal functions

=over 4

=item year

Return just the year element of date.

=item month

Just like year.

=item day

Just like year and month.

=item is_leap

Boolean.  Note that (for B<::Hebrew> and B<::Gregorian>, at least!) this
can be also be used as a static.  That is, you can either say

  $date->is_leap

or

  Date::Convert::Hebrew->is_leap(5757)

=back

=head2 Private functions that are required of all subclasses

You shouldn't call these, but if you want to add a class, you'll need to
write them!  Or it, since at the moment, there's only one.

=over 4

=item initialize

Read in args and initialize object based on their values.  If there are no
args, initialize with the base class's initialize (which will initialize in
the default way described above for B<new>.)  Note the American spelling of
"initialize": "z", not "s".

=back


=head1 SUBCLASS SPECIFIC NOTES

=head2 Absolute

The "Absolute" calendar is just the number of days from a certain reference
point.  Calendar people should recognize it as the "Julian Day Number" with
one minor modification:  When you convert a Gregorian day n to absolute,
you get the JDN of the Gregorian day from noon on.

Since "absolute" has no notion of years it is an extremely easy calendar
for conversion purposes.  I stole the "absolute" calendar format from
Reingold's emacs calendar mode, for debugging purposes.

The subclass is little more than the base class, and as the lowest common
denominator, doesn't have any special functions.

=head2 Gregorian

The Gregorian calendar  is a purely solar calendar, with  a month that
is only  an approximation  of a lunar  month. It is  based on  the old
Julian (Roman)  calendar. This is the  calendar that has been  used by
most of the Western world for the  last few centuries. The time of its
adoption varies  from country  to country. This  B<::Gregorian> allows
you to extrapolate  back to 1 A.D., as per  the programming tradition,
even though the calendar definitely was not in use then.

In addition to the required methods, B<Gregorian> also has B<year>,
B<month>, B<day>, and B<is_leap> methods.  As mentioned above, B<is_leap>
can also be used statically.

=head2 Hebrew

This is the traditional Jewish calendar.  It's based on the solar year, on
the lunar month, and on a number of additional rules created by Rabbis to
make life tough on people who calculate calendars.  :)  If you actually wade
through the source, you should note that the seventh month really does come
before the first month, that's not a bug.

It comes with the following additional methods: B<year>, B<month>, B<day>,
B<is_leap>, B<rosh>, B<part_add>, and B<part_mult>.  B<rosh> returns the
absolute day corresponding to "Rosh HaShana" (New year) for a given year,
and can also be invoked as a static.  B<part_add> and B<part_mult> are
useful functions for Hebrew calendrical calculations are not for much else;
if you're not familiar with the Hebrew calendar, don't worry about them.


=head2 Islamic

The traditional Muslim calendar, a purely lunar calendar with a year that
is a rough approximation of a solar year.  Currently unimplemented.

=head2 Julian

The old Roman calendar, allegedly named for Julius Caesar.  Purely solar,
with a month that is a rough approximation of the lunar month.  Used
extensively in the Western world up to a few centuries ago, then the West
gradually switched over to the more accurate Gregorian.  Now used only by
the Eastern Orthodox Church, AFAIK.


=head1 ADDING NEW SUBCLASSES

This section describes how to extend B<Date::Convert> to add your favorite
date formats.  If you're not interested, feel free to skip it.  :)

There are only three function you I<have> to write to add a new subclass:
you need B<initialize>, B<date>, and B<date_string>.  Of course, helper
functions would probably help. . .  You do I<not> need to write a B<new> or
B<convert> function, since the base class handles them nicely.

First, a quick conceptual overhaul: the base class uses an "absolute day
format" (basically "Julian day format") borrowed from B<emacs>.  This is
just days numbered absolutely from an extremely long time ago.  It's really
easy to use, particularly if you have emacs and emacs' B<calendar mode>.
Each Date::Convert object is a reference to a hash (as in all OO perl) and
includes a special "absol" value stored under a reserved "absol" key.  When
B<initialize> initializes an object, say a Gregorian date, it stores
whatever data it was given in the object and it also calculates the "absol"
equivalent of the date and stores it, too.  If the user converts to another
date, the object is wiped clean of all data except "absol".  Then when the
B<date> method for the new format is called, it calculates the date in the
new format from the "absol" data.

Now that I've thoroughly confused you, here's a more compartmentalized
version:

=over 4

=item initialize

Take the date supplied as argument as appropriate to the format, and
convert it to "absol" format.  Store it as C<$$self{'absol'}>.  You might
also want to store other data, ie. B<::Gregorian> stores C<$$self{'year'}>,
C<$$self{'month'}>, and C<$$self{'day'}>.  If no args are supplied,
explicitly call the base class's initialize,
ie. C<Date::Convert::initialize>, to initialize with a default 'absol' date
and nothing else.

I<NOTE:>  I may move the default behavior into the new constructor.

=item date

Return the date in a appropriate format.  Note that the only fact that
B<date> can take as given is that C<$$self{'absol'}> is defined, i.e. this
object may I<not> have been initialized by the B<initialize> of this
object's class.  For instance, you might have it check if C<$$self{'year'}>
is defined.  If it is, then you have the year component, otherwise, you
calculate year from C<$$self{'absol'}>.

=item date_string

This is the easy part.  Just call date, then return a pretty string based
on the values.

=back


I<NOTE:> The B<::Absolute> subclass is a special case, since it's nearly an
empty subclass (ie. it's just the base class with the required methods
filled out).  Don't use it as an example!  The easiest code to follow would
have been B<::Julian> except that Julian inherits from B<::Gregorian>.
Maybe I'll reverse that. . .


=head1 EXAMPLES

  #!/usr/local/bin/perl5

  use strict;
  use warnings;
  use Date::Convert;

  $date = Date::Convert::Gregorian->new(1974, 11, 27);
  Date::Convert::Hebrew->convert($date);
  print $date->date_string, "\n";

My Gregorian birthday is 27 Nov 1974.  The above prints my Hebrew birthday.

  Date::Convert::Gregorian->convert($date);
  print $date->date_string, "\n";

And that converts it back and prints it in Gregorian.

  $guy = Date::Convert::Hebrew->new(5756, 7, 8);
  print $guy->date_string, " -> ";
  Date::Convert::Gregorian->convert($guy);
  print $guy->date_string, "\n";

Another day, done in reverse.

  @a = (5730, 3, 2);
  @b = Date::Convert::Hebrew->new(@a)->date;
  print "@a\n@b\n";

The above should be an identity for any list @a that represents a
legitimate date.

  #!/usr/local/bin/perl -an

  use Date::Convert;

  $date = Date::Convert::Gregorian->new(@F);
  Date::Convert::Hebrew->convert($date);
  print $date->date_string, "\n";

And that's a quick Greg -> Hebrew conversion program, for those times when
people ask.

=head1 PROBLEMS AND KNOWN ISSUES

The module ignores  the differences between the  definitions of dates.
More  specifically,  Gregorian  and   Julian  dates  are  midnight  to
midnight, Hebrew dates are sunset to sunset and Julian Day Numbers are
noon  to noon.  There is  no  provision to  distinguish between  these
definitions.

Years are  numbered with the astronomers'  and mathematicians' scheme,
that  is,  there  is a  year  zero  preceding  year  one. To  get  the
historians' scheme, you  will have to shift by one  year. For example,
what the  modules call "Year  -45" corresponds to what  the historians
call "46 BC".

=head1 VERSION

Date::Convert 0.17 (pre-alpha)

=head1 ACKNOWLEDGEMENTS AND FURTHER READING

=head2 By MORTY in 1997

The basic idea of using astronomical dates as an intermediary between all
calculations comes from Dershowitz and Reingold.  Reingold's code is the
basis of emacs's calendar mode.  Two papers describing their work (which I
used to own, but lost!  Darn.) are:

``Calendrical Calculations'' by Nachum Dershowitz and Edward M. Reingold,
I<Software--Practice and Experience>, Volume 20, Number 9 (September,
1990), pages 899-928.  ``Calendrical Calculations, Part II: Three
Historical Calendars'' by E. M. Reingold, N. Dershowitz, and S. M. Clamen,
I<Software--Practice and Experience>, Volume 23, Number 4 (April, 1993),
pages 383-404.

They were also scheduled to come out with a book on calendrical
calculations in Dec. 1996, but as of March 1997, it still isn't out yet.

The Hebrew calendrical calculations are largely based on a cute little
English book called I<The Hebrew Calendar> (I think. . .)  in a box
somewhere at my parents' house.  (I'm organized, see!)  I'll have to dig
around next time I'm there to find it.  If you want to access the original
Hebrew sources, let me give you some advice: Hilchos Kiddush HaChodesh in
the Mishneh Torah is not the Rambam's most readable treatment of the
subject.  He later wrote a little pamphlet called "MaAmar HaEibur" which is
both more complete and easier to comprehend.  It's included in "Mich't'vei
HaRambam" (or some such; I've I<got> to visit that house), which was
reprinted just a few years ago.

Steffen Beyer's L<Date::Calc> showed me how to use MakeMaker and write POD
documentation.  Of course, any error is my fault, not his!

=head2 By JFORGET in 2020

Twenty-three years  later, I<Calendrical  Calculations> is  already at
its 4th edition. I have the 3rd edition, published in 2008.

Several alternative calendar Perl modules  have been published, but it
does  not mean  that L<Date::Convert>  is now  irrelevant. After  all,
I<There Is More Than One Way To Do It>.

=head1 SEE ALSO

perl(1)

L<Date::Convert::French_Rev> an example of a new subclass added to the
basic ones.

L<Date::Calc>

L<DateTime>, with L<DateTime::Calendar::Julian> and L<DateTime::Calendar::Hebrew>

L<Date::Converter>

L<Date::Gregorian::Simple>, L<Date::Hebrew::Simple> and L<Date::Julian::Simple>

=head1 AUTHORS

Module creator: Mordechai T. Abzug <morty at umbc dot edu>

Unofficial co-maintainer: Jean Forget <JFORGET at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 1997,  2000, 2020  Mordechai Abzug  and Jean  Forget. All
rights reserved.

This  program  is  free  software. You  can  distribute,  modify,  and
otherwise mangle  Date::Convert under the  same terms as  Perl 5.16.3:
GNU Public License version 1 or later and Perl Artistic License

You can  find the text  of the licenses  in the F<LICENSE> file  or at
L<https://dev.perl.org/licenses/artistic.html>
and L<https://www.gnu.org/licenses/gpl-1.0.html>.

Here is the summary of GPL:

This program is  free software; you can redistribute  it and/or modify
it under the  terms of the GNU General Public  License as published by
the Free  Software Foundation; either  version 1, or (at  your option)
any later version.

This program  is distributed in the  hope that it will  be useful, but
WITHOUT   ANY  WARRANTY;   without  even   the  implied   warranty  of
MERCHANTABILITY  or FITNESS  FOR A  PARTICULAR PURPOSE.   See  the GNU
General Public License for more details.

You  should have received  a copy  of the  GNU General  Public License
along with  this program; if not,  see <https://www.gnu.org/licenses/>
or write to the Free Software Foundation, Inc., L<https://www.fsf.org>.

=cut
